### DISCUSSION TITLE
Unused bindings throw error

### DISCUSSION THREAD

**Discussion by Rus-Abd** [gpuweb/gpuweb#4582] [Q&A] [answered]
```rs
   struct Uniforms { resolution: vec2f, time: f32 };
   @binding(0) @group(0) var<uniform> U : Uniforms;
   @binding(1) @group(0) var mySampler: sampler; 
   @binding(2) @group(0) var myTexture: texture_2d<f32>;
    struct out{
	position: vec4f,
	color: vec3f
    }
	
	 @fragment
	fn main(@builtin(position) fragCoord : vec4f,
					@location(0) color: vec3f) -> @location(0) vec4<f32>
	{
		var uv = fragCoord.xy/U.resolution ;//* 2. -1.;
		//uv*= vec2f(1, -1);
		var col:vec4f =  textureSample(myTexture,mySampler,uv);
		return vec4<f32>(color,1.); 
	}
```
Everything works fine until I comment this line
```rs
var col:vec4f =  textureSample(myTexture,mySampler,uv);
```
I receive this error
```
Number of entries (3) did not match the number of entries (1) specified in [BindGroupLayout].
Expected layout: [{ binding: 0, visibility: ShaderStage::(Vertex|Fragment), buffer: { type: BufferBindingType::Uniform, hasDynamicOffset: 0, minBindingSize: 16 } }]
 - While validating [BindGroupDescriptor] against [BindGroupLayout]
 - While calling [Device].CreateBindGroup([BindGroupDescriptor]).1
 ```
 I'm getting an error for not using these bindings, am I forced to use them?

**Comment by toji** [comment, marked as answer]
I'll point out that this issue list is for problems with the specs, not questions about using the API.

That being said, this is intended WGSL behavior. The workaround is to use a [phony assignment](https://gpuweb.github.io/gpuweb/wgsl/#phony-assignment-section) to indicate to WGSL that the variable is still "in use".

```rs
 @fragment
	fn main(@builtin(position) fragCoord : vec4f,
					@location(0) color: vec3f) -> @location(0) vec4<f32>
	{
		var uv = fragCoord.xy/U.resolution ;//* 2. -1.;
		//uv*= vec2f(1, -1);
                
                // Do a phony assignment to ensure that these variables are not removed from the autogenerated bind group layout.
                _ = myTexture;
                _ = mySampler;
		//var col:vec4f =  textureSample(myTexture,mySampler,uv);
		return vec4<f32>(color,1.); 
	}

```

The other way to work around this is to stop using `layout: 'auto'` when creating your pipelines. By explicitly specifying the `BindGroupLayout` you can consistently create bind groups using every specified entry regardless of if any given shader makes use of all of them.

**Reply by Rus-Abd** [reply]
Sry, didn't rly know where to ask. And thanks for making it clear!

**Reply by toji** [reply]
No worries! I'm glad we were able to help!

**Reply by Johnrobmiller** [reply]
Sorry to rock the boat, but this is not great design. The compiler should be able to infer which variables are used or not and make adjustments accordingly. For example, instead of making the user insert a line like `_ = someBinding`, the compiler could just as easily insert this at runtime.

The workaround, using the "_" character, is not the end of the world, but it's an inelegant way to design to design an API. It's the sort of thing that new users would never know about unless somebody explicitly teaches them, and also the sort of thing that would cause incredible frustration until it is learned.

**Reply by dj2** [reply]
The compiler does infer which ones are used, which is why you have to specify others if they aren't used. We can't just insert fake phony entries for every module scope var because the used vars has to match to the bindings on the API side. You can have multiple entry points in a single file and each of those can use a disjoint set of variables. We don't know which ones you want unless you tell us.

That said, there are discussions on if there should be the option to provide the bindings in a more explicit fashion, but that requires design/review and concensus work. See https://github.com/gpuweb/gpuweb/issues/4957 for that work.
